R/data.table.R

```diff
@@ -1861,90 +1861,243 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
-      else X[[j]] = as.list(as.vector(xj))
+      if (!is.recursive(X[[j]])) { 
+        X[[j]] = as.list(as.vector(X[[j]]))
+      }
     }
-  }
-  else if (all.logical) { }
-  else if (non.numeric) {
-    for (j in seq_len(p)) {
-      if (is.character(X[[j]])) next
-      xj = X[[j]]
-      miss = is.na(xj)
-      xj = if (length(levels(xj))) as.vector(xj) else format(xj)
-      is.na(xj) = miss
-      X[[j]] = xj
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Convert factors to character vectors
+  any.factors = ("factor" %chin% uniq.classes)
+  if (any.factors) {
+    which.factors = which(sapply(col.classes, function(cl_vec) { "factor" %chin% cl_vec }))
+    for (j in which.factors) {
+      miss = is.na(X[[j]])
+      X[[j]] = as.vector(X[[j]])
+      is.na(X[[j]]) = miss
+    }
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Classes to be converted to character vectors
+  charconvert.classes = c("Date", "POSIXct", "POSIXlt", "raw")
+  any.charconvert = any(charconvert.classes %chin% uniq.classes)
+  if (any.charconvert) {
+    which.charconvert = which(sapply(col.classes, function(cl_vec) { any(charconvert.classes %chin% cl_vec) }))
+    for (j in which.charconvert) {
+      miss = is.na(X[[j]])
+      X[[j]] = format(X[[j]])
+      is.na(X[[j]]) = miss
+    }
+    
+    # redetermine classes and types for next checks
+    col.classes = lapply(X, class)
+    uniq.col.classes = unique(col.classes)
+    uniq.classes = unique(unlist(uniq.col.classes))
+    col.types = sapply(X, typeof)
+    uniq.col.types = unique(col.types)
+  }
+  
+  # Other classes from suggested packages that have special type conversion rules
+  any.bit64 = ("integer64" %chin% uniq.classes)
+  if (any.bit64 && length(uniq.classes) > 1L) {
+    # Determine target class we're converting to 
+    target.class = NULL
+    if (all(uniq.classes %chin% c("integer64", "logical", "integer")))
+      target.class = "integer64"
+    else if (all(uniq.classes %chin% c("integer64", "numeric", "complex", "character")))
+      target.class = "character" # integer64 cannot reliably be converted to numeric or complex
+    
+    if (!is.null(target.class)) { # if NULL fallback on type coercion
+      # Which columns need to be coerced?
+      which.convert = which(sapply(col.classes, function(cl_vec) { !any(target.class %chin% cl_vec) }))
+      
+      # Coerce the columns
+      for (j in which.convert) {
+        switch(target.class,
+               "integer64"={ X[[j]] = bit64::as.integer64(X[[j]]) },
```

Comment:
`bit64` is Suggests, need to branch for the case when it's not available I guess

Reply:
My intuition here was that `class(X[[j]]) == "integer64"` should only arise if the bit64 package is installed and loaded, although I suppose someone could load an object saved by another user and they themselves do not have the bit64 package installed.

Reply:
IINM this might happen from `fread`, where we create `bit64` columns even without the pkg loaded IIRC

Reply:
I've added logic to detect this (and set the target class to "character") - we might also consider throwing a warning letter the user know that one or more of their columns is integer64 class but the bit64 package is not installed, so the columns will be coerced to character - if this case ever arises I imagine it would save quite a bit of pain trying to track down why the resulting matrix is a character matrix

Reply:
Changed to error - if bit64 is not installed, then coercing to character will give the wrong result.